articles/setting-up-own-login-vs-login-via-external-services/index.md

```diff
@@ -0,0 +1,68 @@
+* Suppose you get 10k comments on one of the posts on your service, or you get 1k orders within hours of launch on your product on an e-commerce website. These events are updating the database, which, if done without authentication, can lead to one or more of these.
+* Unable to map the comments/orders to the respective user who did it.
+* Partial or complete Denial of Service (DoS). These updates could have been done by a malicious script which in case of limited quantity products in e-commerce(and similar) services render them not be in a state to cater to relevant users.
+
+![Unless secured,the database can be compromised](/engineering-education/setting-up-own-login-vs-login-via-external-services/hacker-stealing-representation.jpg)
+
+### Basic Setup for Login-Signup Flow
+ Login-Signup is the first line of security for any service. If this is breached, a malicious user can access data that is otherwise not allowed. The login interface should also verify for user access level during the process- vendors, delivery boy, buyer, etc. in case of e-commerce or editor, author, a viewer in case of a blog, and similarly for other services. At the bare minimum, the authentication should be on username/email with the password, which can be secured even further with multi-factor authentication. To ensure privacy and protection of the users, passwords must be hashed at all stages(even before sendingit to the server via API).
+
+### Some pointers to keep in mind while hashing are:
+* Use hash function, not encryption, no matter how secure the encryption is. There is no feature or facility required for login and sign-up flow that is not provided by hashing methodology which encryption provides.
+* Use powerful hashing methods with salt(random salt preferably; having the same salt for all passwords makes salting go in vain). Hashing algorithms like MD5 are too weak for present-day computational power, hence comparatively insecure.
+* Additionally, older standards of hashing have multiple known collision pairs. A collision is an event where two distinct inputs give the same hash as output.
+* If you are using API and passing hashed password in request(hashed passwords always better than plain-text), then do not directly store this has in server. If you directly store the hash, then your entire database if compromised in multiple ways.  
+  * Any internal employee with access to the database can copy hash from the database and use API to login.
+  * If the database gets breached by a hacker, they no longer need to get a plain-text password to login as API can do it directly
+
+ Hence, you should hash passwords at each level of transmission, multiple times if needed.
```

Comment:
Multiple rounds of hashing are recommended to increase the computational cost to reverse a hash back to its password in the event of a breach, but this should be done once, at one level, not multiple times.